File: 9/9/list.cpp
#include "list.h"
#include <string>

struct ListElement
{
	std::string word;
	ListElement *next;
};

struct List
{
	int numberOfElements;
	ListElement *head;
};

///Returns an empty list
List *createList()
{
	List *newList = new List;
	newList->numberOfElements = 0;
	newList->head = nullptr;
	return newList;
}

///Adds element to the top of the list
void addElement(List *list, std::string word)
{
	ListElement *listElement = new ListElement;
	listElement->word = word;
	listElement->next = list->head;
	list->head = listElement;
	++list->numberOfElements;
}

///Returns head of the list
ListElement *getHead(List *list)
{
	return list->head;
}

///Returns word in given list element
std::string getWord(ListElement *element)
{
	return element->word;
}

///Returns list element that is located after the given 
ListElement *getNextListElement(ListElement *element)
{
	return element->next;
}

///Deletes list element from the top
void deleteListElement(List *list)
{
	ListElement *temp = list->head;
	list->head = list->head->next;
	delete temp;
	--list->numberOfElements;
}

///Deletes list
void deleteList(List *list)
{
	while (list->head != nullptr)
	{
		deleteListElement(list);
	}
	delete list->head;
}

File: hw4_2/hw2_2/List.cs
namespace List
{
    /// <summary>
    /// List.
    /// </summary>
    class List: IList
    {
        protected ListElement head = null;

        protected class ListElement
        {
            public int Value { get; set; }

            public ListElement Next { get; set; }

            public ListElement()
            {
                Value = 0;
                Next = null;
            }

            public ListElement(int value, ListElement next)
            {
                Value = value;
                Next = next;
            }
        }

        public bool Contains(int element)
        {
            if (!this.IsEmpty())
            {
                ListElement temp = this.head;
                while (temp != null)
                {
                    if (temp.Value == element)
                    {
                        return true;
                    }
                    temp = temp.Next;
                }
            }
            return false;
        }

        public virtual void AddElement(int element)
        {
            if (this.head == null)
            {
                this.head = new ListElement(element, null);
            }
            else
            {
                ListElement temp = new ListElement(element, this.head);
                this.head = temp;
            }
        }

        public bool IsEmpty() =>  this.head == null;

        public void DeleteElement(int element)
        {
            if (this.IsEmpty())
            {
                throw new EmptyListException();
            }

            ListElement temp = this.head;
            if (temp.Value == element)
            {
                this.head = this.head.Next;
                return;
            }
            if (temp.Next != null)
            {
                if (temp.Next.Value == element)
                {
                    this.head.Next = this.head.Next.Next;
                    return;
                }
            }
            else
            {
                throw new NonexistentElementException();
            }

            while (temp.Next != null)
            {
                if (temp.Next.Value == element)
                {
                    temp.Next = temp.Next.Next;
                    return;
                }
                temp = temp.Next;
            }
            throw new NonexistentElementException();
        }
    }
}